core/src/main/java/com/linecorp/armeria/client/ClientFactory.java

```diff
@@ -96,12 +103,38 @@ static ClientFactoryBuilder builder() {
      * Closes the default {@link ClientFactory}.
      */
     static void closeDefault() {
-        LoggerFactory.getLogger(ClientFactory.class).debug(
-                "Closing the default {}", ClientFactory.class.getSimpleName());
-        try {
-            DefaultClientFactory.DEFAULT.doClose();
-        } finally {
-            DefaultClientFactory.INSECURE.doClose();
+        final Logger logger = LoggerFactory.getLogger(ClientFactory.class);
+        logger.debug("Closing the default client factories");
+        final CompletableFuture<Void> closeFuture = CompletableFuture.allOf(
+                DefaultClientFactory.DEFAULT.closeAsync(false),
+                DefaultClientFactory.INSECURE.closeAsync(false)).handle((unused1, cause) -> {
+            if (cause == null) {
+                logger.debug("Closed the default client factories");
+            } else {
+                logger.warn("Failed to close the default client factories:", Exceptions.peel(cause));
+            }
+            return null;
+        });
+
+        if (!(Thread.currentThread() instanceof NonBlocking)) {
+            boolean interrupted = false;
+            try {
+                for (;;) {
```

Comment:
Can you add a comment why we don't use `join()` here? I think this is basically only called from a shutdown hook ever, so interruption probably doesn't matter so much but not sure.

Reply:
It's because it may not be invoked from a shutdown hook? Nothing stops a user from calling this method manually (or automatically via some framework).